topics: Apply automatic follow/unmute when message moved to new topic.

roanster007 opened the pull request:
This commit adds a feature, wherein when someone moves a user's message to be the first message in a topic, and the user has the Automatically follow topics initiated enabled, then the new topic will also be followed by the user.

Similarly, if the user has Automatically unmute topics initiated enabled, the moved topic would also be unmuted.

This is done through a new function -
apply_automatic_unmute_follow_topics_policy.

Fixes #28408

<!-- Describe your pull request here.-->


<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

![ZIuQYReZFi](https://github.com/zulip/zulip/assets/97049524/aa098a22-0dff-4a1d-9a1e-873f972e0184)


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

roanster007 commented on zerver/actions/message_edit.py:
```diff
@@ -1003,6 +1031,13 @@ def user_info(um: UserMessage) -> Dict[str, Any]:
+        if first_message_moved:
+            apply_automatic_unmute_follow_topics_policy(
+                user_profile, first_message_moved, target_stream, target_topic_name
+            )
```
Does it require a comment ?

roanster007 commented on zerver/actions/message_edit.py:
```diff
@@ -1003,6 +1031,13 @@ def user_info(um: UserMessage) -> Dict[str, Any]:
+        if first_message_moved:
+            apply_automatic_unmute_follow_topics_policy(
+                user_profile, first_message_moved, target_stream, target_topic_name
+            )
```
I didn't make an if condition to verify if it is a bot or not before applying `apply_automatic_unmute_follow_topics_policy`, assuming no one would move a bot's message to another topic. Is it needed?

timabbott replied:
That assumption doesn't make any sense, of course any message can be moved.

timabbott commented on zerver/actions/message_edit.py:
```diff
@@ -385,6 +385,34 @@ def get_visibility_policy_after_merge(
+            target_stream,
+            target_topic_name,
+            visibility_policy=UserTopic.VisibilityPolicy.UNMUTED,
+        )
```
This logic does not seem to match the spec -- there's no check for how the user's existing policy relates to the new one.

roanster007 replied:
updated!

prakhar1144 replied:
I don't think we need `select_related` here. Also, we should avoid `.get()` pattern as [we discussed earlier](https://github.com/zulip/zulip/pull/28980#discussion_r1502084484).

This can be simplified to:
```
Subscription.objects.filter(recipient=target_stream.recipient, user_profile=user_profile,
active=True, is_user_active=True)
```

Note: `active=True` is important here, otherwise a user who had unsubscribed the stream will have an entry in UserTopic table -- which is not correct. (Maybe this will raise error in this case -- haven't tested manually).

prakhar1144 commented on zerver/actions/message_edit.py:
```diff
@@ -385,6 +385,34 @@ def get_visibility_policy_after_merge(
 
+def apply_automatic_unmute_follow_topics_policy(
+    user_profile: UserProfile,
+    message: Message,
```
* We are not using the `user_profile` parameter in the function at all.
* Also, we are using only the `sender`, so we can pass that directly instead of `message`.

prakhar1144 commented on zerver/actions/message_edit.py:
```diff
@@ -1003,6 +1031,13 @@ def user_info(um: UserMessage) -> Dict[str, Any]:
                     visibility_policy=new_visibility_policy,
                 )
 
+    if topic_name is not None and not target_topic_has_messages:
```
This doesn't work if we are changing the stream only & topic name is kept the same.

prakhar1144 commented on zerver/actions/message_edit.py:
```diff
@@ -1003,6 +1031,13 @@ def user_info(um: UserMessage) -> Dict[str, Any]:
                 )
 
+    if topic_name is not None and not target_topic_has_messages:
+        first_message_moved = changed_messages.first()
```
I think we can directly use `target_message` here ? -- This will avoid the extra DB query too.

roanster007 replied:
if we use `target_message` directly, I guess there would be a small bug -- when you select "move message" for some message which is not the first message of the topic, and you choose the option to move all messages of topic, then the `automatic follow` is applied to this `target_message` on which we selected move, rather than the first message of the topic.

roanster007 replied:
I updated it according to your point.

roanster007 commented:
puppeteer tests failing at download

prakhar1144 commented:
I spent an hour thinking about few edge cases:
* What if the message is moved to a stream which the user can't access -- the current code creates a user topic row (that shouldn't happen)
* If all the messages are moved (which is a common case when some new user creates a topic in unrelated stream), we are already handling that behaviour (just above the current code block). So the current code will lead to duplicate DB queries & logs. -- Note that we are not considering "automatic follow/unmute policy", so we need to make that tweak, if required.
* If the user had has a visibility policy set for old topic, whether the "automatic follow/unmute policy" should prevail ?

prakhar1144 commented on zerver/actions/message_edit.py:
```diff
@@ -385,6 +386,37 @@ def get_visibility_policy_after_merge(
+        user_profile.automatically_unmute_topics_in_muted_streams_policy
+        == UserProfile.AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_ON_INITIATION
+    ):
+        subscription = Subscription.objects.get(
```
We should avoid using `.get()` -- see: https://zulip.readthedocs.io/en/latest/contributing/code-style.html#never-do-direct-database-queries-userprofile-objects-get-client-objects-get-etc

roanster007 replied:
thanks for letting me know! I'll update it.

roanster007 commented:
> * If the user had has a visibility policy set for old topic, whether the "automatic follow/unmute policy" should prevail ?
I guess you're right... if `propagate_mode == "change_all` then the visibility policy set for the previous topic should prevail (that would make more sense).... in that case, we could also reduce a db query.

I updated it.

roanster007 commented:
@prakhar1144 could you have another look at this, since tim and alya agreed upon the "automatic follow/unmute" not being applied in case when visibility policy is `change_all` in the [#design](https://chat.zulip.org/#narrow/stream/101-design/topic/moving.20topic.20to.20first.20message/near/1749049) thread.

prakhar1144 commented:
@roanster007 thanks for the ping. I'll take a look in the morning tomorrow.

prakhar1144 commented on zerver/tests/test_message_edit.py:
```diff
@@ -1596,3 +1598,100 @@ def test_user_group_mention_restrictions_while_editing(self) -> None:
+            UserProfile.AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_ON_INITIATION,
+            acting_user=None,
+        )
+        result = self.client_patch(
```
We can use `do_edit_message_assert_success`

prakhar1144 replied:
Hm, that's right. I'll let @timabbott comment on this if it is worth the effort.

timabbott replied:
The explicit API call seems totally fine to me.

prakhar1144 commented on zerver/tests/test_message_edit.py:
```diff
@@ -1596,3 +1598,100 @@ def test_user_group_mention_restrictions_while_editing(self) -> None:
+            stream_id=stream.id,
+            topic_name="testing",
+        )
+        user_ids = stream_topic_target.user_ids_with_visibility_policy(
```
I think the right pattern to test here would be:

* assert that user was not following the topic
* do the move topic operation
* assert that user now follows.

The reason to do that first step is to avoid a case where the topic is already followed (due to some reason) -- then the test after that would be of no use for us. :)

roanster007 replied:
updated

roanster007 commented on zerver/actions/message_edit.py:
```diff
@@ -1003,6 +1035,11 @@ def user_info(um: UserMessage) -> Dict[str, Any]:
 
+    elif (topic_name is not None or new_stream is not None) and not target_topic_has_messages:
+        sender = target_message.sender
+        if not sender.is_bot and sender.id not in losing_access_user_ids:
```
oh, thanks for letting me know. I guess we can still use `users_losing_access`?

prakhar1144 replied:
Yeah, actually we are still using these IDs in `send_event(user_profile.realm, delete_event, [user.id for user in users_losing_access])`. So, we can just restore `users_losing_access_ids` and use it at both here & our use case.

roanster007 replied:
rather than that I guess I could directly use `sender not in  users_losing_access`.

prakhar1144 commented on zerver/actions/message_edit.py:
```diff
@@ -387,6 +387,42 @@ def get_visibility_policy_after_merge(
+            user_profile=user_profile,
+            active=True,
+            is_user_active=True,
+            is_muted=True,
```
Let's not include `is_muted` in query otherwise we won't be benefiting with the partial index we created - see:
```
    class Meta:
        unique_together = ("user_profile", "recipient")
        indexes = [
            models.Index(
                fields=("recipient", "user_profile"),
                name="zerver_subscription_recipient_id_user_profile_id_idx",
                condition=Q(active=True, is_user_active=True),
            ),
        ]
```

roanster007 replied:
ohk...I'll make sure to take care of indexes from now on before making db calls. thanks!

prakhar1144 commented:
LGTM.

@zulipbot add "integration review"

roanster007 commented:
@timabbott I addressed all the reviews of Prakhar, leaving [this](https://github.com/zulip/zulip/pull/28980#discussion_r1535016522), for which he has asked your views:

>We can use do_edit_message_assert_success

> I don't think we can... that method is defined only inside the test -- test_edit_message_content_limit . Maybe do you want me to do a prep commit to move the definition of that method outside this test to be able to use it in other tests?

> Hm, that's right. I'll let timabbott comment on this if it is worth the effort.

roanster007 commented:
This is ready for another review.

timabbott commented on zerver/tests/test_message_edit.py:
```diff
@@ -1655,3 +1657,114 @@ def test_user_group_mention_restrictions_while_editing(self) -> None:
+            stream_name="new_stream",
+            topic_name="test",
+            content="Always here",
+        )
```
A few comments on the test:

- Let's use more semantic names for your topics. `test` and `testing` don't tell the reader anything about the intent. Names like `original` and `post_move` would be better, though see below, because you're likely to . You could also think about just having a variable with a name like `original_topic` rather than being expressive with the string topic names.
- I don't understand why the `Shiva` user gets referenced at all. Was the intent to have them move topics, via `self.api_patch` so you don't have to switch who is logged in, but you later realized Hamlet had the permission? Though maybe you do need this user, see the next note.
- This test doesn't verify the full set of logic/conditions for this code. It's important to verify both that (1) a user whose message is moved to be the first message in a new topic has their visibility policy changed but also (2) a user who sent a message that is moved to be a not-first message in a new topic, or into an existing topic, does not have their visibility policy changed. Your tests don't attempt to verify the various corner cases around that -- like what happens if you move an older message to be the new first message in an existing topic? The tests should verify that whatever happens in that case is as intended.

Likely to address this you'll need to extend these tests to do more setup and/or make multiple moves. So, for example, you could have:
- Iago sends a message to "original"
- Cordelia sends a message to "original"
- Hamlet sends a message to "original"
- Shiva sends a message to "original"
- Confirm only Iago follows the topic "original".
- Move Hamlet&Shiva messages to "post-move"
- Confirm Hamlet now follows "post-move"
- Move Cordelia message to "post-move", which notably is now the first message in the topic.
- Confirm cordelia does now follow "post-move" too.

timabbott replied:
You probably also need some form of test coverage for the private stream with protected history case.

roanster007 replied:
updated!

timabbott commented on zerver/actions/message_edit.py:
```diff
@@ -1005,6 +1040,11 @@ def user_info(um: UserMessage) -> Dict[str, Any]:
+    elif (topic_name is not None or new_stream is not None) and not target_topic_has_messages:
+        sender = target_message.sender
+        if not sender.is_bot and sender not in users_losing_access:
+            apply_automatic_unmute_follow_topics_policy(sender, target_stream, target_topic_name)
```
I think there's two things wrong with just checking `target_topic_has_messages`.
- The first is that if the stream is a private stream with protected history, we need to be checking not whether your message is now the very first message in the topic, but whether there are any visible-to-you messages in the topic. (We can't have any logic producing results visible to you depend on whether there are message in a topic on an absolute basis, or you use that to determine whether a topic exists that you don't have access to). I think that means we need to instead compute `bulk_access_stream_messages_query(sender, messages_for_topic(realm_id, new_stream.recipient_id, target_topic_name), new_stream).first()`, and compare its ID to `target_message.id`.
- This logic, as written, doesn't consider the possibility that one moves a message to be the new first message in a topic that already has messages. I hint at this a bit [here](https://github.com/zulip/zulip/pull/28980/files#r1651791539), but I discussed with @alya and she agrees that in this case, we should have the user follow the topic, to preserve the invariant that if you see a topic where user X is the first message, you can probably assume they're following that topic and you don't need to mention them, regardless of what set of moves might have gotten them there. I think the proposed logic in the previous bullet fixes this too, but wanted to flag it because we'll want the tests to verify it too.

Probably doing that access check will be more expensive/slow than would be ideal, so we probably need to also try to keep track of whether we already moved an existing followed-topic record in the `user_profiles_having_visibility_policy` logic above. This is important, because as written, if you just rename an existing topic, it's very likely that we'd end up doing a bunch of useless database queries to try to mark the first poster in that topic as following it twice within the same `do_update_messages` function call.

roanster007 replied:
>The first is that if the stream is a private stream with protected history, we need to be checking not whether your message is now the very first message in the topic, but whether there are any visible-to-you messages in the topic.

I guess this is being taken care by `users_losing_access`:
```py
# If it's moving to a private stream, all non-subscribed users are losing access
 users_losing_access = users_losing_usermessages
```

roanster007 replied:
>This logic, as written, doesn't consider the possibility that one moves a message to be the new first message in a topic that already has messages.

Makes sense, I'll update it.

roanster007 replied:
>if you just rename an existing topic, it's very likely that we'd end up doing a bunch of useless database queries to try to mark the first poster in that topic as following it twice within the same do_update_messages function call.

renaming a topic would mean `moved_all_visible_messages` will be `True`, which won't be executed under this condition as per the [CZO thread](https://chat.zulip.org/#narrow/stream/101-design/topic/moving.20topic.20to.20first.20message/near/1749049):
>We already handle renaming entire topics by moving visibility policies, so I think the goal would be for this logic to not fire in that case.

timabbott commented:
I posted a couple comments on issues that are blockers for merging this. This implementation should work fine in the common cases, but we do need to be correct in corner cases too, which the tests don't attempt to verify, and after thinking about how to verify them in the tests, I realized that the logic is in fact wrong in those corner cases.

timabbott commented:
Dropping the integration review label since this has bugs that need to be resolved as the next step.

roanster007 commented:
Addressed all reviews!

alya commented:
@timabbott I added the "integration review" label, since your feedback has been addressed. Please feel free to redirect if something else should happen first.

whilstsomebody commented:
Hey @roanster007 , can I continue with your work on the issue?

timabbott commented:
I think it makes sense for @roanster007 to rebase as the next step on this.

timabbott commented:
The updated version looks great (though please remember to post a top-level PR comment when ready for a next review! It's very helpful to have the status always visible here). I made these changes to save a useless DB query and clarify the test.

``` diff
diff --git a/zerver/actions/message_edit.py b/zerver/actions/message_edit.py
index cfe5a34af3..d033b3692b 100644
--- a/zerver/actions/message_edit.py
+++ b/zerver/actions/message_edit.py
@@ -1199,15 +1199,15 @@ def do_update_message(
             realm.id, target_stream.recipient_id, target_topic
         ).exclude(id__in=[*changed_message_ids])
 
-        first_message_in_target_topic = None
-
-        # We need to filter out messages which are not accessible to the
-        # sender -- such as in cases when channel is protected and there were
-        # messages before the user joined.
-        if messages_in_target_topic is not None:
-            first_message_in_target_topic = bulk_access_stream_messages_query(
-                sender, messages_in_target_topic, target_stream
-            ).first()
+        # All behavior in channels with protected history depends on
+        # the permissions of users who might glean information about
+        # whether the topic previously existed. So we need to look at
+        # whether this message is becoming the first message in the
+        # target topic, as seen by the user who we might change topic
+        # visibility policy for in this code path.
+        first_message_in_target_topic = bulk_access_stream_messages_query(
+            sender, messages_in_target_topic, target_stream
+        ).first()
 
         is_target_message_first = False
         # the target_message would be the first message in the moved topic
diff --git a/zerver/tests/test_message_edit.py b/zerver/tests/test_message_edit.py
index baad4c0d5a..f679449448 100644
--- a/zerver/tests/test_message_edit.py
+++ b/zerver/tests/test_message_edit.py
@@ -2577,6 +2577,7 @@ class EditMessageTest(ZulipTestCase):
             acting_user=None,
         )
 
+        # Iago's message to "test topic" is not visible to Hamlet.
         core = self.make_stream("core", iago.realm, True, history_public_to_subscribers=False)
         self.subscribe(iago, "core")
         self.send_stream_message(iago, "core", topic_name="test topic")
@@ -2591,6 +2592,16 @@ class EditMessageTest(ZulipTestCase):
             topic_name="test topic",
         )
 
+        self.assertEqual(
+            stream_topic_target_post_move.user_ids_with_visibility_policy(
+                UserTopic.VisibilityPolicy.FOLLOWED
+            ),
+            set(),
+        )
+
+        # Verify that Hamlet follows the topic after moving his
+        # message, because as far as he knows, his message is now the
+        # first message in "test topic".
         self.check_automatic_change_visibility_policy_on_initiation_during_moving_messages(
             msg_id,
             hamlet.id,
@@ -2598,3 +2609,10 @@ class EditMessageTest(ZulipTestCase):
             UserTopic.VisibilityPolicy.FOLLOWED,
             True,
         )
+
+        self.assertEqual(
+            stream_topic_target_post_move.user_ids_with_visibility_policy(
+                UserTopic.VisibilityPolicy.FOLLOWED
+            ),
+            set([hamlet.id]),
+        )
```

With those changes, merged, thanks @roanster007!
